File: Assets/Plugins/Android/AndroidFeatures.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace KlimSoft
{
    public static class AndroidFeatures
    {
        private static AndroidJavaClass featuresClass;
        private static bool initialized = false;

        public static int DialogButtonId
        {
            get
            {
                if (!initialized)
                    Initialize();

                int id = featuresClass.GetStatic<int>("dialogButtonId");
                featuresClass.SetStatic<int>("dialogButtonId", -1);
                return id;
            }
        }

        private static void Initialize()
        {
            featuresClass = new AndroidJavaClass("KlimSoft.AndroidFeatures");

            //Get context and activity.
            AndroidJavaClass unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
            AndroidJavaObject activity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity");
            AndroidJavaObject context = activity.Call<AndroidJavaObject>("getApplicationContext");

            //Set context and activity.
            featuresClass.CallStatic("setContext", context);
            featuresClass.CallStatic("setActivity", activity);

            initialized = true;
        }

        public static void Vibrate(long milliseconds)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("vibrate", milliseconds);
#endif
        }

        public static void Vibrate(long[] intervals, int repeat = -1)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("vibrate", intervals, repeat);
#endif
        }

        public static void MakeToast(string text, int length = 0)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("makeToast", text, length);
#endif
            Debug.Log("Computer case 2 toast: " + text);
        }

        public static void KillProcess()
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("killProcess");
#endif
        }

        public static string GetSignature()
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            return featuresClass.CallStatic<string>("getSignature");
#else
            return null;
#endif
        }

        public static void CheckSignature(string trueSignature)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("checkSignature", trueSignature);
#endif
        }

        public static void ShowDialogWithOneButton(string message, string title, string buttonText = "Ok", bool cancelable = true)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            featuresClass.CallStatic("showDialogWithOneButton", message, title, buttonText, cancelable);
#endif
        }

        public static string[] GetPackageNames()
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            return featuresClass.CallStatic<string[]>("getPackageNames");
#else
            return null;
#endif
        }

        public static bool IsAppInstalled(string[] triggers)
        {
#if !UNITY_EDITOR && UNITY_ANDROID
            if (!initialized)
                Initialize();

            bool res = false;
            try
            {
                res = featuresClass.CallStatic<bool>("isAppInstalled", new object[] { triggers });
            }
            catch (Exception e)
            {
                e.ToString();
            }
            return res;
#else
            return false;
#endif
        }
    }
}

File: Assets/Plugins/Android/AndroidFeatures.java
package KlimSoft;

import android.content.Context;
import android.os.Vibrator;
import android.os.VibrationEffect;
import android.widget.Toast;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.app.Activity;
import android.content.pm.ApplicationInfo;
import java.util.List;

class AndroidFeatures
{
	private static Context context;
	private static Activity activity;
	private static Vibrator vibrator;

	public static int dialogButtonId = -1;

	public static void setContext(Context contextParam)
	{
		context = contextParam;
	}

	public static void setActivity(Activity activityParam)
	{
		activity = activityParam;
	}

	public static int getApiLevel()
	{
		return android.os.Build.VERSION.SDK_INT;
	}

	public static void vibrate(long milliseconds)
	{
		if (context == null)
			throw new NullPointerException("Context is not assigned.");

		try
		{
			if (vibrator == null)
				vibrator = (Vibrator)context.getSystemService("vibrator");

			if (getApiLevel() >= 26)
			{
				VibrationEffect effect = VibrationEffect.createOneShot(milliseconds, -1);
				vibrator.vibrate(effect);
			}
			else
			{
				vibrator.vibrate(milliseconds);
			}
		}
		catch(Exception e) { };
	}

	public static void vibrate(long[] intervals, int repeat)
	{
		if (context == null)
			throw new NullPointerException("Context is not assigned.");

		try
		{
			if (vibrator == null)
				vibrator = (Vibrator)context.getSystemService("vibrator");

			if (getApiLevel() >= 26)
			{
				VibrationEffect effect = VibrationEffect.createWaveform(intervals, repeat);
				vibrator.vibrate(effect);
			}
			else
			{
				vibrator.vibrate(intervals, repeat);
			}
		}
		catch(Exception e) { };
	}

	public static void makeToast(String text, int length)
	{
		if (context == null)
			throw new NullPointerException("Context is not assigned.");

		Toast.makeText(context, text, length).show();
	}

	public static void killProcess()
	{
		android.os.Process.killProcess(android.os.Process.myPid());
	}

	public static String getSignature()
	{
		if (context == null)
			throw new NullPointerException("Context is not assigned.");

		try
		{
			Signature[] signatures = context
			.getPackageManager()
			.getPackageInfo(context.getPackageName(), PackageManager.GET_SIGNATURES)
			.signatures;

			return signatures[0].toCharsString();
		}
		catch (Exception e)
		{
			return null;
		}
	}

	public static void checkSignature(String trueSignature)
	{
		String currentSignature = getSignature();

		if (currentSignature == null)
			killProcess();

		if (currentSignature != trueSignature)
			killProcess();
	}

	public static void showDialogWithOneButton(String message, String title, String buttonText, boolean cancelable)
	{
		if (activity == null)
			throw new NullPointerException("Activity is not assigned.");

		AlertDialog.Builder builder = new AlertDialog.Builder(activity);

		builder.setMessage(message)
		.setTitle(title)
		.setCancelable(cancelable)
		.setNeutralButton(buttonText, new DialogInterface.OnClickListener()
		{
			public void onClick(DialogInterface dialog, int id)
			{
				dialogButtonId = id;
			}
		});

		builder.create().show();
	}

	public static String[] getPackageNames()
	{
		if (context == null)
			throw new NullPointerException("Context is not assigned.");

		PackageManager packageManager = context.getPackageManager();
		List<ApplicationInfo> packages = packageManager.getInstalledApplications(PackageManager.GET_META_DATA);
		String[] result = new String[packages.size()];

		for (int i = 0; i < packages.size(); i++)
			result[i] = packages.get(i).packageName;

		return result;
	}

	public static boolean isAppInstalled(String[] triggers)
	{
		String[] packageNames = getPackageNames();

		for (int i = 0; i < triggers.length; i++)
			for (int j = 0; j < packageNames.length; j++)
				if (packageNames[j] == triggers[i])
					return true;
		
		return false;
	}
}